### PR TITLE
compiler: Patch premature lowering of EvalDerivative

### LINKED ISSUE #1978: Premature lowering of EvalDerivative
Reproducer:

```
        grid = Grid(shape=(10, 10))

        u = TimeFunction(name='u', grid=grid, space_order=4)

        # Not really a custom derivative, but the enforced pre-evaluation makes
        # it behaves as if it were one
        mock_custom_deriv = u.dx.dy.evaluate

        # This symbolic operation -- creating an Add between an arbitray object
        # and an EvalDerivative -- caused the EvalDerivative to be prematurely
        # simplified being flatten into an Add
        expr0 = u.dt - mock_custom_deriv
        expr = -expr0

        eq = Eq(u.forward, expr)

        op = Operator(eq)
```

flop count blows up here because `EvalDerivatives` get lost during the early symbolic manipulations

### PR DESCRIPTION
**Pull request by FabioLuporini**
fixes #1978

### CODE REVIEW

**Review by georgebisbas — APPROVED** [review]

File: devito/finite_differences/differentiable.py [hunk 1]
```diff
@@ -423,14 +423,30 @@ def __new__(cls, *args, **kwargs):
         # a*1 -> a
         args = [i for i in args if i != 1]
 
+        # a*-1 -> a
```
Comment (mloubout) [inline devito/finite_differences/differentiable.py:426]: -a ?

File: devito/finite_differences/differentiable.py [hunk 2]
```diff
@@ -423,14 +423,30 @@ def __new__(cls, *args, **kwargs):
...
         # a*-1*-1 -> a
+        # a*-1*-1*-1 -> a*-1
```
Comment (mloubout) [inline devito/finite_differences/differentiable.py:428]: -a ?

File: devito/finite_differences/differentiable.py [hunk 3]
```diff
@@ -423,14 +423,30 @@ def __new__(cls, *args, **kwargs):
...
         # Reorder for homogeneity with pure SymPy types
         _mulsort(args)
 
+        # `sympy.Mul.flatten(coeff, Add)` flattens out nested Adds within Add,
+        # which would destroy `EvalDerivative`s if present. So here we perform
+        # a similar thing, but cautiously construct an evaluated Add, which
+        # will preserve the integrity of `EvalDerivative`s, if any.
+        try:
+            a, b = args
+            if not a.is_zero and a.is_Rational:
```
Comment (mloubout) [inline devito/finite_differences/differentiable.py:443]: `a.is_zero` is un-necessary would have been cought line 420 already.

File: devito/finite_differences/differentiable.py [hunk 4]
```diff
@@ -423,14 +423,30 @@ def __new__(cls, *args, **kwargs):
...
+                r, b = b.as_coeff_Mul()
+                if r is sympy.S.One and type(b) is Add:
```
Comment (mloubout) [inline devito/finite_differences/differentiable.py:445]: Can this be generalized to like `r is sympy.Number` or `r.is_Symbol` and merge it with `a`?

**Review by mloubout — COMMENTED** [review]
Some comments but nice catch and yest these `1.0` where a pain

File: devito/finite_differences/differentiable.py (hunk 4 above)
Comment (FabioLuporini) [reply devito/finite_differences/differentiable.py:445]: for the record, I lifted it from sympy

https://github.com/sympy/sympy/blob/master/sympy/core/mul.py#L285

tho admittedly a slightly older version than the one I have installed locally

I will try to improve it

File: devito/finite_differences/differentiable.py (hunk 2 above)
Comment (FabioLuporini) [reply devito/finite_differences/differentiable.py:428]: the arguments are `a` and `-1` (unless `a` it's an integer) that's why it's not factored out

File: devito/finite_differences/differentiable.py (hunk 1 above)
Comment (FabioLuporini) [reply devito/finite_differences/differentiable.py:426]: good catch. will put `a*-1` (see below)

**Comment by FabioLuporini** [comment]
pushed edits

**Comment by FabioLuporini** [comment]
note latest pep8-related commit necessary due to https://github.com/devitocodes/devito/runs/7605926638?check_suite_focus=true

**Review by EdCaunt — APPROVED** [review]

**Review by mloubout — APPROVED** [review]

**Pull request merged**